src/program.ts

```diff
@@ -1062,6 +1067,19 @@ export class Program extends DiagnosticEmitter {
             );
             queuedImports.splice(i, 1);
             madeProgress = true;
+
+            // Also, make sure to record the namespace in our importNamespaces for later
+            if (importNamespacesMap.has(foreignFile)) {
+              let importNamespaceLocalFileMap = importNamespacesMap.get(foreignFile) as Map<File, Array<string>>;
+              let importNamespaceNames = importNamespaceLocalFileMap.get(localFile) as Array<string>;
```

Comment:
This line can `abort`s if `localFile` is not yet in the map, needs another `has` check.

In general the mechanism here appears reasonable, but I'd suspect that it can be simplified a bit. For instance, we reference the imported namespace above via `foreignFile.asImportedNamespace` and when we have references there might be an opportunity to link related imported namespaces together so whenever we add to one of them we can add to all of them. Like `Namespace#aliasNamespaces: Set<Namespace>` and upon `Namespace#add` add to aliases as well? Just an idea so far, didn't dive deep :)

Reply:
> This line can aborts if localFile is not yet in the map, needs another has check.

Ah yeah, I totally need to add the `has` check, was a bas assumption on my part, nice catch! :smile: 

> In general the mechanism here appears reasonable, but I'd suspect that it can be simplified a bit. For instance, we reference the imported namespace above via foreignFile.asImportedNamespace and when we have references there might be an opportunity to link related imported namespaces together so whenever we add to one of them we can add to all of them. Like Namespace#aliasNamespaces: Set<Namespace> and upon Namespace#add add to aliases as well? Just an idea so far, didn't dive deep :)

Ah, so to be honest, I'm a little lost in that explanation :upside_down_face: But I do agree it could be done better! I was already having a hard time understanding what was going on, so I was more trying to just at least not make a hack haha! :joy: But if I'm not mistaken, it seems like you want to change the map from:

foreignFile -> localFile -> NamespaceNames

to 

foreignFile -> localFile

And then the localFile would then introduce something to know "Hey, for this foreignfile, these namespaces of mine need to be updated"? :thinking:

Reply:
Yeah, that's all a bit confusing. What I'm suggesting is that when we convert a file to a namespace in `File#asImportedNamespace`, we can keep track of all the namespaces we made from that same file, and implicity add new elements to all the converted namespaces when adding an element to the file in `File#add`. Let's say we have a `Set` `File#asImportedNamespaces` containing all the namespaces we made from that file, then instead of adding everything manually, adding any element to that `File` would also populate all the `asImportedNamespaces` automatically, in turn greatly simplifying this.

Reply:
Ahhhhhh! That makes perfect sense, cool! I'll do that now! :smile: